Add mystnb format

chrisjsewell opened the pull request:
Heya @mwouts the new format we have discussed for you to look at 😄 

closes #447, originally implemented in ExecutableBookProject/MyST-NB#82

See https://myst-parser.readthedocs.io and https://myst-nb.readthedocs.io for further details on the format. The file extension is still up for discussion. Also, at present, I copy `nbformat` and `nbformat_minor` to the front matter, but that's probably not necessary?

All tests pass, apart from 3 that are skipped due to diffs in newline characters at the end of markdown blocks, which don't actually affect the consistency of the round-trip.

cc'ing @choldgraf @AakashGfude @mmcky @jstac

chrisjsewell commented:
Hmm, all tests are passing for the required python versions, but `conda create -q -n jupytext-env --file requirements-dev.txt --file requirements.txt;`
doesn't like `myst-parser>=0.7.1,<0.8; python_version >= '3.6'`.
Sure you don't want to just drop python 2.7 & 3.4 in this PR 😉

chrisjsewell commented on .travis.yml:
```diff
@@ -39,7 +39,7 @@ install:
     fi
   # install is required for testing the pre-commit mode
-  - pip install . || true
+  - pip install .[myst] || true
```
I moved the install to an extra, if thats ok

chrisjsewell commented:
@choldgraf you'll like this one; in d627398 I have introduced logic to compact the metadata blocks.

For blocks with no nested dicts, the block is denoted by starting colons::

````
```{nb-code}
:other: true
:tags: [hide-output, show-input]

content
```
````

For blocks with nesting the block is enlosed by ``---``::

````
```{nb-code}
---
other:
  more: true
tags: [hide-output, show-input]
---

content
```
````

mwouts commented:
Hello @chrisjsewell , that's awesome!

Sorry to have caused you trouble with Python 2.7, I prefer to keep testing it, but I agree, it's hard to make the CI work with conditional dependencies. Adding the extra dependencies in the CI script itself is perfectly OK. Anyway, maybe I should rewrite the part of `travis.yaml` that tests conda environment to match the recommendations from [CONTRIBUTING.md](https://github.com/mwouts/jupytext/blob/master/CONTRIBUTING.md#how-to-setup-a-development-environment-for-jupytext)...

Tonight I'll take the time to review this with more detail. We could ship that soon, at least if it's OK on your side. Maybe we could 
- confirm the `mystnb` extension
- add a word/link on the `myst` format in the documentation (in `formats.md`)
- extend the Notebook and Lab extension with the `myst` format

Do you see anything else that you'd like to add?

chrisjsewell commented:
> to match the recommendations from CONTRIBUTING.md.

Oh yeh,  should have probably read that lol!

> confirm the mystnb extension

I've added `.mnb` now as well as `.mystnb`. I'm happy with these, but @choldgraf, @mmcky can interject if they wish.

> - add a word/link on the myst format in the documentation (in formats.md)
> - extend the Notebook and Lab extension with the myst format
>
>    Do you see anything else that you'd like to add?

No, I think that should just about cover it 👍

choldgraf commented on tests/notebooks/mirror/ipynb_to_myst/Notebook with function and cell metadata 164.mystnb:
````diff
@@ -0,0 +1,42 @@
+A markdown cell
+And below, the cell for function f has non trivial cell metadata. And the next cell as well.
+
+```{nb-code} ipython3
````
Isn't this planned to be `{execute}`?

choldgraf commented:
This looks great to me - with one main question, which is that I see a few discrepancies between the `.myst` format and the format we've laid out in:

https://github.com/ExecutableBookProject/MyST-NB/issues/12#issue-567866971

In particular, using `{nb-code}` instead of `{execute}`, same for `{nb-raw}` etc (where I assumed we would just use `{raw}`.

Is there a reason for the change?

chrisjsewell commented:
> In particular, using `{nb-code}` instead of `{execute}`, same for `{nb-raw}` etc (where I assumed we would just use `{raw}`.

This is the 'tricky' part; currently `{execute}` is a synonym for `{jupyter-execute}` and implies that the directive will be able to be run by the [`jupyter_sphinx.JupyterCell`](https://github.com/jupyter/jupyter-sphinx/blob/970097e01c1b7af04372be202aed572de68d2d20/jupyter_sphinx/ast.py#L26) directive, if you just have the `myst-parser` and `jupyter-sphinx` extensions in your sphinx `conf.py`.
The following will **not** work in this way, because the metadata/options are invalid for `JupyterCell`.

````
```{execute}
:other: true
:tags: [hide-output, show-input]

content
```
````

If you wanted this to be directly used by `JupyterCell` you would have to introduce a specific conversion of the metadata to comply with the options dictated by that directive. This would negate true round-trip conversion, because you would have to 'throw away' any metadata that `JupyterCell` doesn't have an option for.

> Another option just be to use `.md` like many of the other parsers do, since that's such an overloaded extension anyway...

This then follows on from the above; with this "`.mystnb`" file format, we are saying that it will not be treated the same as simple `.md` in the sphinx conversion; it will be specifically converted to a notebook first (either using a pre-conversion step in [NotebookParser](https://github.com/ExecutableBookProject/MyST-NB/blob/4c3f36b884086878dc0af5dfcfd735b98ebbbacf/myst_nb/parser.py#L24) or a separate sub-class of it [as I outlined here](https://github.com/ExecutableBookProject/MyST-NB/pull/82#issuecomment-599254868)). So in your sphinx project you may have three distinct types of file:

- `.md` which are read directly by the myst-parser package's [MystParser](https://github.com/ExecutableBookProject/MyST-Parser/blob/62e1bf9a95fa683b4c4e2aaac1ad4fd94b855ce6/myst_parser/sphinx_parser.py#L12). Which could contain `jupyter-execute` directives, **but would not** be involved in any jupyter-cache like execution caching. 
- `.ipynb` which are read by the myst-nb package's [NotebookParser](https://github.com/ExecutableBookProject/MyST-NB/blob/4c3f36b884086878dc0af5dfcfd735b98ebbbacf/myst_nb/parser.py#L24)
- `.mystnb` which are read by `NotebookParser` or a `MystNbParser` sub-class.

In this context, `.ipynb` and `.mystnb` are interchangeble.

chrisjsewell commented:
In 4c206ec I have added MyST-NB to the Notebook extension. Basically, anywhere there was an `Rmd` I also added an `mnd`. From manual testing this looks to be working great, but obviously @mwouts may want to check over this

chrisjsewell commented:
FYI I had a few difficulties following the `jupytext/nbextension/README.md`, for which this issue helped maartenbreddels/ipyvolume#85

chrisjsewell commented:
I added MyST to the jupyterlab extension, but I'm having trouble installing it to test. So I've put the commit it in a separate branch for now a6a95e8 might need you to add that for me @mwouts 

FYI I note that in the package.json, the jupyterlab dependancy is set to version 1, but now version 2 is released. I'm guessing that's on the TODO list to update

choldgraf commented:
That all sounds good to me @chrisjsewell - @mwouts if we end up changing the names of some things then we are happy to upstream a PR to keep jupytext up-to-date with things.

chrisjsewell commented:
Added mystnb to benchmarking (4c206ec) and no problems there

<img width="600" alt="image" src="https://user-images.githubusercontent.com/2997570/76803339-7e1ee080-67d1-11ea-993b-69b35fc66513.png">

<img width="600" alt="image" src="https://user-images.githubusercontent.com/2997570/76803355-8aa33900-67d1-11ea-90ce-192d77257e8c.png">

mwouts commented:
Impressive work @chrisjsewell !

I'll merge this now, and then maybe we can do a bit more testing, and ship this in a few days...

A few comments:
- This PR seems to be based on version 1.3.4, now we are at 1.4.0+dev, but I expect no issue with the merge.
- I can take care of adding the `mnb` extension to the JupyterLab extension, sure no pb
- Maybe I'll regenerate the `World population.mystnb` example as a paired notebook (e.g. ` jupytext --set-formats ipynb,.pct.py:percent,.lgt.py:light,.spx.py:sphinx,md,Rmd,.pandoc.md:pandoc,mnb,mystnb '.\World population.ipynb'`) - I'll do that after the merge
- We will also have to add a mention of the `mnb` and `mystnb` extensions in the documentation

chrisjsewell commented:
> We will also have to add a mention of the mnb and mystnb extensions in the documentation

In a few more minutes I would of committed that 😆 

> This PR seems to be based on version 1.3.4, now we are at 1.4.0+dev, but I expect no issue with the merge

Ah oops, that must have been down to way I fork it off of the ExecutableBookProject:master

> I can take care of adding the mnb extension to the JupyterLab extension, sure no pb
> Maybe I'll regenerate the World population.mystnb example as a paired notebook (e.g. jupytext --set-formats ipynb,.pct.py:percent,.lgt.py:light,.spx.py:sphinx,md,Rmd,.pandoc.md:pandoc,mnb,mystnb '.\World population.ipynb') - I'll do that after the merge

Cheers 😄

mwouts commented:
> I added MyST to the jupyterlab extension, but I'm having trouble installing it to test. So I've put the commit it in a separate branch for now [a6a95e8](https://github.com/mwouts/jupytext/commit/a6a95e833d74ad5116640c5361adcbf0ff496e6d) might need you to add that for me @mwouts
> 
> FYI I note that in the package.json, the jupyterlab dependancy is set to version 1, but now version 2 is released. I'm guessing that's on the TODO list to update

@chrisjsewell , if you like you're actually very close to having a PR for this... your change is correct, but should be rebased to match the latest version of the Lab extension (and version number could thus be 1.2.1). Just one detail - could you also update `packages/labextension/CHANGELOG.md` before compiling the `.tar.gz` package?

And if you want to try the lab extension locally, I recommend building (`python setup.py sdist bdist_wheel`) and then pip install the `jupytext-1.4.0+dev.tar.gz` package, cf. the [extension README.md](https://github.com/mwouts/jupytext/blob/master/packages/labextension/README.md).
